Add interface for defining external statistics.

BenKaufmann opened the pull request:
Dear Max,

I created an alternative implementation for your pull request #20 that is a little bit more in the spirit of the existing clasp code. 

Please let me know if this interface suits your needs or if you have any comments. Given that this new interface is quite low-level and not type-safe, I deliberately used more explicit function names, like e.g. `mapAdd()` and `arrayAdd()`.

rkaminsk commented:
I'll talk to Max. It is probably a good idea to try to use this for the statistics in clingo and give feedback based on this. -R

pluehne commented on CHANGES:
```diff
@@ -1,5 +1,6 @@
 clasp 3.3.4:
   * fixed: Atoms of incremental programs not always marked as frozen.
+  * added internal interface for user defined statistics
```
For consistency, I’d write “user-defined“ with a hyphen here as well.

BenKaufmann replied:
Yep. Thanks.

pluehne commented on src/clasp_facade.cpp:
```diff
@@ -696,13 +725,17 @@ ClaspFacade::Statistics::ClingoView::ClingoView(const ClaspFacade& f) {
 }
 void ClaspFacade::Statistics::ClingoView::update(const ClaspFacade::Statistics& stats) {
+	if (stats.userStats_.get() && !stats.userStats_->empty()) {
+		keys_.add("userdefined", stats.userStats_->toStats());
```
Is there a reason why `userdefined` is spelled without a hyphen here? I know that comment is a bit picky, but it might be inconvenient to make such things consistent at a later point in time.

BenKaufmann replied:
For consistency, this should be `user_defined`. Or alternatively, `external` to indicate that these are not maintained by clasp.

pluehne replied:
Okay. I was just referring to the fact that user-defined is a compound adjective and should be written with a hyphen, but if hyphens are not allowed within keys, that’s just how it is :slightly_smiling_face:.

BenKaufmann replied:
Sorry, I forgot to add the "Yep, you are right". I just wanted to express that keys typically use underscores (and unfortunately sometimes camel-case) instead of hyphens.

pluehne replied:
I totally understand, thanks for the info … and sorry about the fuss and my pedantry anyway :smile:!

MaxOstrowski commented:
Unused array indices are filled with empty objects.
Example:
https://github.com/MaxOstrowski/clasp/tree/external_stats
Is it possible to make them of type empty, as the clingo interface supports this statistics_type?

MaxOstrowski commented:
Would it be possible to add a little more functionality for inspecting the values (iteration)
e.g:
mapSize()
arraySize()
mapAt(index)
Then we could unify the two statistic interfaces to one with the same functionality (except that original clasp statistics can not be altered)

Thanks a lot Benni

BenKaufmann commented:
@MaxOstrowski I addressed the two issues you raised, i.e. I unified the interface (this also includes a new branch in libpotassco) and changed the behaviour of `add()` for arrays.

- Regarding array `add()`: holes are now filled with objects of type `Empty`, which can later be overwritten. I do not particular like this change because previously `Empty` was only used to signal errors (in places where exceptions are undesirable). Maybe we could switch to a less general `push(...)` interface that only allows for appending single objects thereby preventing the creation of holes in the first place.
- For integrating the stats interfaces, I introduced a new `writable` property indicating whether calling add on a key is allowed. Starting from a writable root, new writable keys can be added but all existing clasp stats are considered read-only.

I'm not sure whether I got the interplay between the "user root" and the stats callbacks right. The current implementation simply calls the callbacks with a stats object that is rooted at "user_stats". That is, inside of callbacks, clasp statistics are hidden and new stats automatically occur under "user_stats".
However, alternative designs are certainly possible:

- We could pass the full stats object (including the clasp stats) and either add a new parameter `userRoot` of type `Key_t` to the callbacks or add a function `userRoot()` to the `AbstractStatistics` interface.
- We could get rid of the notion of a dedicated user root and instead let callbacks add their own root keys.  To handle output in clasp, we could then either limit printing to some dedicated key (e.g. "user_stats") or go the whole nine yards and print all user roots.

Any preferences?

MaxOstrowski commented:
@BenKaufmann Thanks a lot for all your efforts.
I talked to Roland and he liked the idea of the push interface, which should be enough for handling all use cases.
The writeable flag is good for the clasp interface. You understood the intention of user_root correctly.
The plan is to make the statistics object const in the old interface and non-const for the callbacks.
Also user_root should be used during the callbacks.

> I'm not sure whether I got the interplay between the "user root" and the stats callbacks right. The current implementation simply calls the callbacks with a stats object that is rooted at "user_stats". That is, inside of callbacks, clasp statistics are hidden and new stats automatically occur under "user_stats".

This is perfectly fine

BenKaufmann commented:
@MaxOstrowski I implemented the `push()` based array interface and updated the PR accordingly. If you are happy with the current version, I'll merge the PR to dev. Otherwise, let me know what is missing.

MaxOstrowski commented:
Looks nice, thank you very much.
Roland requested a has_subkey function that I'm not able to implement.
Would it be possible to add such a thing ?

MaxOstrowski commented:
Exactly. This function allows Roland to implement any clingo interface he wants. Thanks a lot.

BenKaufmann commented:
@rkaminsk I understand your point but just to clarify: The suggested `tryGet()` will also have linear complexity and at least for non-writable keys it will even have to apply a `try/catch`-based search, where non-existence is internally communicated via an exception. So, `tryGet()` will have ok performance in the best and pretty horrible performance in the worst case (if we assume a table-based approach to exceptions).

BenKaufmann commented:
@rkaminsk @MaxOstrowski I added `bool AbstractStatistics::find(Key_t mapK, const char* key, Key_t* outKey)`. Anything else missing or should I merge to dev?

MaxOstrowski commented:
You can merge with dev, erverything should work now.
Thanks a lot !

rkaminsk commented:
Thanks, was too slow to agree too. :wink:
